File: src/main/java/com/senderman/lastkatkabot/command/user/Stats.java
package com.senderman.lastkatkabot.command.user;

import com.annimon.tgbotsmodule.api.methods.Methods;
import com.annimon.tgbotsmodule.commands.context.MessageContext;
import com.senderman.lastkatkabot.command.CommandExecutor;
import com.senderman.lastkatkabot.dbservice.UserStatsService;
import com.senderman.lastkatkabot.util.Html;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.objects.User;

@Component
public class Stats implements CommandExecutor {

    private final UserStatsService users;


    public Stats(UserStatsService users) {
        this.users = users;
    }

    @Override
    public String command() {
        return "/stats";
    }

    @Override
    public String getDescription() {
        return "статистика. Реплаем можно узнать статистику реплайнутого";
    }

    @Override
    public void accept(MessageContext ctx) {
        User user = (ctx.message().isReply()) ? ctx.message().getReplyToMessage().getFrom() : ctx.user();

        if (user.getIsBot()) {
            ctx.replyToMessage("Но это же просто бот, имитация человека! " +
                               "Разве может бот написать симфонию, иметь статистику, играть в BnC, участвовать в дуэлях?")
                    .callAsync(ctx.sender);
            return;
        }

        var stats = users.findById(user.getId());
        String name = Html.htmlSafe(user.getFirstName());
        int winRate = stats.getDuelsTotal() == 0 ? 0 : 100 * stats.getDuelWins() / stats.getDuelsTotal();
        String text = String.format(
                "📊 Статистика %s:\n\n" +
                "👑 Дуэлей выиграно: %d\n" +
                "⚔️ Всего дуэлей: %d\n" +
                "📈 Винрейт: %d\n\n" +
                "🐮 Баллов за быки и коровы: %d",
                name, stats.getDuelWins(), stats.getDuelsTotal(), winRate, stats.getBncScore());

        if (stats.getLoverId() != null) {
            var lover = Methods.getChatMember(stats.getLoverId(), stats.getLoverId()).call(ctx.sender);
            if (lover != null) {
                String loverLink = Html.getUserLink(lover.getUser());
                text += "\n\n❤️ Вторая половинка: " + loverLink;
            }
        }
        ctx.reply(text).callAsync(ctx.sender);
    }
}

File: src/main/kotlin/com/senderman/lastkatkabot/usercommands/Stats.kt
package com.senderman.lastkatkabot.usercommands

import com.annimon.tgbotsmodule.api.methods.Methods
import com.senderman.lastkatkabot.LastkatkaBotHandler
import com.senderman.lastkatkabot.Services
import com.senderman.neblib.CommandExecutor
import com.senderman.neblib.TgUser
import org.telegram.telegrambots.meta.api.objects.Message

class Stats(private val handler: LastkatkaBotHandler) : CommandExecutor {
    override val command: String
        get() = "/stats"
    override val desc: String
        get() = "статистика. Реплаем можно узнать статистику реплайнутого"

    override fun execute(message: Message) {
        val player = if (!message.isReply) message.from else message.replyToMessage.from
        if (player.bot) {
            handler.sendMessage(
                message.chatId, "Но это же просто бот, имитация человека! " +
                        "Разве может бот написать симфонию, иметь статистику, играть в BnC, любить?"
            )
            return
        }
        val user = TgUser(player)
        val stats = Services.db.getStats(player.id)
        val (_, duelWins, totalDuels, bnc, loverId, childId, coins) = stats
        val winRate = if (totalDuels == 0) 0 else 100 * duelWins / totalDuels
        var text = """
            📊 Статистика ${user.name}:

            Дуэлей выиграно: $duelWins
            Всего дуэлей: $totalDuels
            Винрейт: $winRate%
            
            💰 Деньги: $coins
            🐮 Баллов за быки и коровы: $bnc
        """.trimIndent()

        text += connectUserToStats(message, loverId, AnotherUser.LOVER)
        text += connectUserToStats(message, childId, AnotherUser.CHILD)
        text += "\n\nХочешь больше монеток? Играй в @miniroulette_bot!"
        handler.sendMessage(message.chatId, text)
    }

    private enum class AnotherUser(val title: String) {
        LOVER("\n❤️ Вторая половинка: "),
        CHILD("\n\uD83D\uDC76\uD83C\uDFFB️ Ребенок: ")
    }

    /**
     * Add line about user connected to userId
     * @param message - message object of /stats caller
     * @param userId - ID of connected user
     * @param type - type of connected user
     * @return empty string if userId == 0, else type.title + link to user if message is pm, or user'name if not pm
     */

    private fun connectUserToStats(message: Message, userId: Int, type: AnotherUser): String {
        if (userId == 0)
            return ""

        val user: TgUser =
            try {
                TgUser(Methods.getChatMember(userId.toLong(), userId).call(handler).user)
            } catch (e: Exception) {
                try {
                    TgUser(Methods.getChatMember(message.chatId, userId).call(handler).user)
                } catch (e: Exception) {
                    TgUser(userId, "Без имени")
                }
            }

        return type.title + if (message.isUserMessage) user.link else user.name
    }
}